lib/Parse/ParseDecl.cpp

```diff
@@ -799,6 +799,269 @@ Parser::parseImplementsAttribute(SourceLoc AtLoc, SourceLoc Loc) {
+        break;
+      }
+      case tok::integer_literal: {
+        unsigned paramNum;
+        if (parseUnsignedInteger(
+                paramNum, paramLoc,
+                diag::diff_params_clause_expected_parameter))
+          return true;
+
+        params.push_back(ParsedAutoDiffParameter::getOrderedParameter(
+            paramLoc, paramNum));
+        break;
+      }
+      case tok::kw_self: {
+        paramLoc = consumeToken(tok::kw_self);
+        params.push_back(ParsedAutoDiffParameter::getSelfParameter(paramLoc));
+        break;
+      }
+      default:
+        diagnose(Tok, diag::diff_params_clause_expected_parameter);
+        return true;
+    }
+    if (parseTrailingComma && Tok.isNot(tok::r_paren))
+      return parseToken(tok::comma, diag::attr_expected_comma, attrName,
+                        /*isDeclModifier=*/false);
+    return false;
+  };
+
+  // Parse opening '(' of the parameter list.
+  if (Tok.is(tok::l_paren)) {
+    SyntaxParsingContext DiffParamsContext(
+        SyntaxContext, SyntaxKind::DifferentiationParams);
+    consumeToken(tok::l_paren);
+    // Parse first parameter. At least one is required.
+    if (parseParam())
+      return errorAndSkipToEnd(2);
+    // Parse remaining parameters until ')'.
+    while (Tok.isNot(tok::r_paren))
+      if (parseParam())
+        return errorAndSkipToEnd(2);
+    SyntaxContext->collectNodesInPlace(SyntaxKind::DifferentiationParamList);
+    // Parse closing ')' of the parameter list.
+    consumeToken(tok::r_paren);
+  }
+  // If no opening '(' for parameter list, parse a single parameter.
+  else {
+    if (parseParam(/*parseTrailingComma*/ false))
+      return errorAndSkipToEnd();
+  }
+  return false;
+}
+
+bool Parser::parseDifferentiableAttributeArguments(
+    bool &linear, SmallVectorImpl<ParsedAutoDiffParameter> &params,
+    Optional<DeclNameWithLoc> &jvpSpec, Optional<DeclNameWithLoc> &vjpSpec,
+    TrailingWhereClause *&whereClause) {
+  StringRef AttrName = "differentiable";
+
+  // Set parse error, skip until ')' and parse it.
+  auto errorAndSkipToEnd = [&](int parenDepth = 1) -> bool {
+    for (int i = 0; i < parenDepth; i++) {
+      skipUntil(tok::r_paren);
+      if (!consumeIf(tok::r_paren))
+        diagnose(Tok, diag::attr_expected_rparen, AttrName,
+                 /*DeclModifier=*/false);
+    }
+    return true;
+  };
+
+  // Parse trailing comma, if it exists, and check for errors.
+  auto consumeIfTrailingComma = [&]() -> bool {
+    if (!consumeIf(tok::comma)) return false;
+    // Diagnose trailing comma before 'where' or ')'.
+    if (Tok.is(tok::kw_where) || Tok.is(tok::r_paren)) {
+      diagnose(Tok, diag::unexpected_separator, ",");
+      return true;
+    }
+    // Check that token after comma is 'wrt:' or a function specifier label.
+    if (!Tok.is(tok::identifier) || !(Tok.getText() == "wrt" ||
+                                      Tok.getText() == "jvp" ||
+                                      Tok.getText() == "vjp")) {
```

Comment:
Would it help to have a function like:

enum { wrt, jvp, vjp, invalid } classifyLabel(StringRef str);

function to keep all the classification logic in sync, and move the magic string parsing into a single place?

Reply:
“jvp:” and “vjp:” will go away very soon with the almost-finished retroactive derivative registration feature (‘@differentiating’). So I think it is fine to leave them as is!

Reply:
Good to know, but it is important to keep master consistent and following best practices.  If these aren't important, then it would be fine to remove them from the patch.  If they need to be in the patch, then please consider implementing them in a nice way :).

I'm not saying that classifyLabel is an appropriate thing, just saying that the rationale doesn't make sense.

Reply:
Ok, sounds good!

Reply:
I have added a bunch of helpers. PTAL.

(Although other code in master is using `Tok.getText()` to compare against literal strings.)